### PR TITLE
Do not restrict the kind of type in a property pattern

### LINKED ISSUE #10158: IDE crash in pattern-matching with array type
reported by @MadsTorgersen 

Typing open curly after array type in pattern.

I guess I won’t be getting the Length out that way… ;-)

Mads

image

### PR DESCRIPTION
**Pull request by gafter**
Fixes #10158 
@AlekseyTs @agocke Please review
@dotnet/roslyn-compiler FYI

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs [hunk 1]
```diff
@@ -11931,5 +11931,31 @@ public class Odd<T>
...
+    }
+    static void M(params string[] args)
+    {
+        if (args is object[] { Length is var length }) { Console.Write(length); }
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlib(source, options: TestOptions.ReleaseExe, parseOptions: patternParseOptions);
+            compilation.VerifyDiagnostics();
+            var verifier = CompileAndVerify(compilation, expectedOutput: @"30");
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: Is there any way the behavior could be observed using the semantic model? If so I would also expect a test there.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs [hunk 2]
```diff
@@ -11931,5 +11931,31 @@ public class Odd<T>
 False
 False");
         }
+
+        [Fact, WorkItem(10158, "https://github.com/dotnet/roslyn/issues/10158")]
+
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: An Unnecessary empty line.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: Done.

Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: @agocke Done. Anything else before you approve?

**Comment by agocke** [comment]
@gafter LGTM

**Pull request merged**